### ISSUE TITLE
dataset page, clicking a  file row checkbox produces 300-400 queries

### ISSUE THREAD

**Issue by raprasad** [IQSS/dataverse#2633]
**location**: dataset page with file listing
**version**: 4.2

### Desired functionality
  - Selecting file checkboxes should enable or disable buttons such as "Download" 
  - The server does not need to be called to accomplish this task

### What happens

On the 4.2 branch (including current production):
  - Clicking a checkbox associated with a file produces:
     -   **300 to 400** queries 
     -  Takes about 1 second of time.  
  - Clicking the 'select all' checkbox produces about the same number of queries--not any more than a single checkbox.

### Implications
  - Poor user experience for people clicking multiple checkboxes
  - Un-needed server load (300-400 queries) and delayed response time
     -  This is bad for a page--this functionality doesn't require any queries

<img width="1055" alt="replication_data_for__a_rhythm_landscape_approach_to_the_developmental_dynamics_of_birdsong_rhythm_-_harvard_dataverse" src="https://cloud.githubusercontent.com/assets/1036060/10416517/4785b21e-6fe8-11e5-98f1-265eb515df85.png">

**Comment by bencomp** [comment]
Say what? 300-400 queries? What kind of queries?

**Comment by raprasad** [comment]
@bencomp 
They are database queries.  Here are the queries when a checkbox is clicked and the user is _not logged in__

```sql
(1) COUNT: 32   QUERY(truncated): SELECT t0.ID, t0.DTYPE, t0.CREATEDATE, t0.INDEXTIME, t0.MODIFICATIONTIME, t0.PERMISSIONINDEXTIME, t0.PERMISSIONMODIFICATIONTIME, t0.PUBLICATIONDATE, t0.CREATOR_ID, t0.OWNER_ID, t0.RELEASEUSER_ID, t1.ID, t1.AFFILIATION, t1.ALIAS, t1.DATAVERSETYPE, t1.description, t1.FACETROOT, t1.GUESTBOOKROOT, t1.METADATABLOCKROOT, t1.NAME, t1.PERMISSIONROOT, t1.TEMPLATEROOT, t1.THEMEROOT, t1.DEFAULTCONTRIBUTORROLE_ID, t1.DEFAULTTEMPLATE_ID FROM DVOBJECT t0, DATAVERSE t1 WHERE ((t0.ID = ?) AND ((t1.ID = t0.ID) AND (t0.DTYPE = ?)))
----------------------------------------
(2) COUNT: 32   QUERY(truncated): SELECT ID, ARCHIVEDESCRIPTION, ARCHIVEURL, HARVESTSTYLE, HARVESTTYPE, HARVESTINGSET, HARVESTINGURL, dataverse_id FROM HARVESTINGDATAVERSECONFIG WHERE (dataverse_id = ?)
----------------------------------------
(3) COUNT: 32   QUERY(truncated): SELECT ID, BACKGROUNDCOLOR, LINKCOLOR, LINKURL, LOGO, LOGOALIGNMENT, LOGOBACKGROUNDCOLOR, LOGOFORMAT, TAGLINE, TEXTCOLOR, dataverse_id FROM DATAVERSETHEME WHERE (dataverse_id = ?)
----------------------------------------
(4) COUNT: 32   QUERY(truncated): SELECT DISTINCT DTYPE FROM DVOBJECT WHERE (ID = ?)
----------------------------------------
(5) COUNT: 30   QUERY(truncated): SELECT ID, AFFILIATION, EMAIL, FIRSTNAME, LASTNAME, MODIFICATIONTIME, POSITION, SUPERUSER, USERIDENTIFIER FROM AUTHENTICATEDUSER WHERE (ID = ?)
----------------------------------------
(6) COUNT: 30   QUERY(truncated): SELECT ID, AUTHENTICATIONPROVIDERID, PERSISTENTUSERID, AUTHENTICATEDUSER_ID FROM AUTHENTICATEDUSERLOOKUP WHERE (AUTHENTICATEDUSER_ID = ?)
----------------------------------------
(7) COUNT: 18   QUERY(truncated): SELECT ID, ALIAS, DESCRIPTION, NAME, PERMISSIONBITS, OWNER_ID FROM DATAVERSEROLE WHERE (ID = ?)
----------------------------------------
(8) COUNT: 14   QUERY(truncated): SELECT t0.ID, t0.DTYPE, t0.CREATEDATE, t0.INDEXTIME, t0.MODIFICATIONTIME, t0.PERMISSIONINDEXTIME, t0.PERMISSIONMODIFICATIONTIME, t0.PUBLICATIONDATE, t0.CREATOR_ID, t0.OWNER_ID, t0.RELEASEUSER_ID, t1.ID, t1.AUTHORITY, t1.DOISEPARATOR, t1.FILEACCESSREQUEST, t1.GLOBALIDCREATETIME, t1.IDENTIFIER, t1.PROTOCOL, t1.guestbook_id, t1.thumbnailfile_id FROM DVOBJECT t0, DATASET t1 WHERE ((t0.ID = ?) AND ((t1.ID = t0.ID) AND (t0.DTYPE = ?)))
----------------------------------------
(9) COUNT: 14   QUERY(truncated): SELECT ID, INFO, STARTTIME, USER_ID, DATASET_ID FROM DATASETLOCK WHERE (DATASET_ID = ?)
----------------------------------------
(10) COUNT: 8   QUERY(truncated): SELECT ID, ASSIGNEEIDENTIFIER, DEFINITIONPOINT_ID, ROLE_ID FROM ROLEASSIGNMENT WHERE ((ASSIGNEEIDENTIFIER = ?) AND (DEFINITIONPOINT_ID = ?))
----------------------------------------
(11) COUNT: 6   QUERY(truncated): SELECT t1.ID, t1.DESCRIPTION, t1.DISPLAYNAME, t1.GROUPALIAS, t1.GROUPALIASINOWNER, t1.OWNER_ID FROM AUTHENTICATEDUSER t0, EXPLICITGROUP_AUTHENTICATEDUSER t2, EXPLICITGROUP t1 WHERE (((t1.OWNER_ID = ?) AND (t0.ID = ?)) AND ((t2.ExplicitGroup_ID = t1.ID) AND (t0.ID = t2.containedAuthenticatedUsers_ID)))
----------------------------------------
(12) COUNT: 4   QUERY(truncated): SELECT NAME, CONTENT FROM SETTING WHERE (NAME = ?)
----------------------------------------
(13) COUNT: 1   QUERY(truncated): SELECT DISTINCT t0.ID, t0.DTYPE, t0.DESCRIPTION, t0.DISPLAYNAME, t0.PERSISTEDGROUPALIAS FROM IPV4RANGE t1 LEFT OUTER JOIN PERSISTEDGLOBALGROUP t0 ON (t0.ID = t1.OWNER_ID) WHERE ((t1.BOTTOMASLONG <= ?) AND (t1.TOPASLONG >= ?))
```

**Comment by raprasad** [comment]
@bencomp 
- These are database queries triggered by selecting/unselecting a checkbox.  
- Below are two examples:
  - from a user who is NOT logged in (64 queries) and
  - logged in user (400+ queries)

Here are the queries for a user who is NOT **logged in**:
```sql
----------------------------------------
(1) COUNT: 32   QUERY(truncated): SELECT NAME, CONTENT FROM SETTING WHERE (NAME = ?)
----------------------------------------
(2) COUNT: 16   QUERY(truncated): SELECT ID, ASSIGNEEIDENTIFIER, DEFINITIONPOINT_ID, ROLE_ID FROM ROLEASSIGNMENT WHERE ((ASSIGNEEIDENTIFIER = ?) AND (DEFINITIONPOINT_ID = ?))
----------------------------------------
(3) COUNT: 12   QUERY(truncated): SELECT t0.ID, t0.DESCRIPTION, t0.DISPLAYNAME, t0.GROUPALIAS, t0.GROUPALIASINOWNER, t0.OWNER_ID FROM EXPLICITGROUP t0, ExplicitGroup_CONTAINEDROLEASSIGNEES t1 WHERE (((t0.OWNER_ID = ?) AND (t1.CONTAINEDROLEASSIGNEES = ?)) AND (t1.ExplicitGroup_ID = t0.ID))
----------------------------------------
(4) COUNT: 4   QUERY(truncated): SELECT DISTINCT t0.ID, t0.DTYPE, t0.DESCRIPTION, t0.DISPLAYNAME, t0.PERSISTEDGROUPALIAS FROM IPV4RANGE t1 LEFT OUTER JOIN PERSISTEDGLOBALGROUP t0 ON (t0.ID = t1.OWNER_ID) WHERE ((t1.BOTTOMASLONG <= ?) AND (t1.TOPASLONG >= ?))
```

**Comment by raprasad** [comment]
424 Queries for a user who is  **logged in**:

```sql
----------------------------------------
(1) COUNT: 48   QUERY(truncated): SELECT t0.ID, t0.DTYPE, t0.CREATEDATE, t0.INDEXTIME, t0.MODIFICATIONTIME, t0.PERMISSIONINDEXTIME, t0.PERMISSIONMODIFICATIONTIME, t0.PUBLICATIONDATE, t0.CREATOR_ID, t0.OWNER_ID, t0.RELEASEUSER_ID, t1.ID, t1.AFFILIATION, t1.ALIAS, t1.DATAVERSETYPE, t1.description, t1.FACETROOT, t1.GUESTBOOKROOT, t1.METADATABLOCKROOT, t1.NAME, t1.PERMISSIONROOT, t1.TEMPLATEROOT, t1.THEMEROOT, t1.DEFAULTCONTRIBUTORROLE_ID, t1.DEFAULTTEMPLATE_ID FROM DVOBJECT t0, DATAVERSE t1 WHERE ((t0.ID = ?) AND ((t1.ID = t0.ID) AND (t0.DTYPE = ?)))
----------------------------------------
(2) COUNT: 48   QUERY(truncated): SELECT ID, ARCHIVEDESCRIPTION, ARCHIVEURL, HARVESTSTYLE, HARVESTTYPE, HARVESTINGSET, HARVESTINGURL, dataverse_id FROM HARVESTINGDATAVERSECONFIG WHERE (dataverse_id = ?)
----------------------------------------
(3) COUNT: 48   QUERY(truncated): SELECT ID, BACKGROUNDCOLOR, LINKCOLOR, LINKURL, LOGO, LOGOALIGNMENT, LOGOBACKGROUNDCOLOR, LOGOFORMAT, TAGLINE, TEXTCOLOR, dataverse_id FROM DATAVERSETHEME WHERE (dataverse_id = ?)
----------------------------------------
(4) COUNT: 48   QUERY(truncated): SELECT DISTINCT DTYPE FROM DVOBJECT WHERE (ID = ?)
----------------------------------------
(5) COUNT: 40   QUERY(truncated): SELECT ID, AFFILIATION, EMAIL, FIRSTNAME, LASTNAME, MODIFICATIONTIME, POSITION, SUPERUSER, USERIDENTIFIER FROM AUTHENTICATEDUSER WHERE (ID = ?)
----------------------------------------
(6) COUNT: 40   QUERY(truncated): SELECT ID, AUTHENTICATIONPROVIDERID, PERSISTENTUSERID, AUTHENTICATEDUSER_ID FROM AUTHENTICATEDUSERLOOKUP WHERE (AUTHENTICATEDUSER_ID = ?)
----------------------------------------
(7) COUNT: 32   QUERY(truncated): SELECT ID, ASSIGNEEIDENTIFIER, DEFINITIONPOINT_ID, ROLE_ID FROM ROLEASSIGNMENT WHERE ((ASSIGNEEIDENTIFIER = ?) AND (DEFINITIONPOINT_ID = ?))
----------------------------------------
(8) COUNT: 32   QUERY(truncated): SELECT ID, ALIAS, DESCRIPTION, NAME, PERMISSIONBITS, OWNER_ID FROM DATAVERSEROLE WHERE (ID = ?)
----------------------------------------
(9) COUNT: 28   QUERY(truncated): SELECT NAME, CONTENT FROM SETTING WHERE (NAME = ?)
----------------------------------------
(10) COUNT: 24   QUERY(truncated): SELECT t1.ID, t1.DESCRIPTION, t1.DISPLAYNAME, t1.GROUPALIAS, t1.GROUPALIASINOWNER, t1.OWNER_ID FROM AUTHENTICATEDUSER t0, EXPLICITGROUP_AUTHENTICATEDUSER t2, EXPLICITGROUP t1 WHERE (((t1.OWNER_ID = ?) AND (t0.ID = ?)) AND ((t2.ExplicitGroup_ID = t1.ID) AND (t0.ID = t2.containedAuthenticatedUsers_ID)))
----------------------------------------
(11) COUNT: 16   QUERY(truncated): SELECT t0.ID, t0.DTYPE, t0.CREATEDATE, t0.INDEXTIME, t0.MODIFICATIONTIME, t0.PERMISSIONINDEXTIME, t0.PERMISSIONMODIFICATIONTIME, t0.PUBLICATIONDATE, t0.CREATOR_ID, t0.OWNER_ID, t0.RELEASEUSER_ID, t1.ID, t1.AUTHORITY, t1.DOISEPARATOR, t1.FILEACCESSREQUEST, t1.GLOBALIDCREATETIME, t1.IDENTIFIER, t1.PROTOCOL, t1.guestbook_id, t1.thumbnailfile_id FROM DVOBJECT t0, DATASET t1 WHERE ((t0.ID = ?) AND ((t1.ID = t0.ID) AND (t0.DTYPE = ?)))
----------------------------------------
(12) COUNT: 16   QUERY(truncated): SELECT ID, INFO, STARTTIME, USER_ID, DATASET_ID FROM DATASETLOCK WHERE (DATASET_ID = ?)
----------------------------------------
(13) COUNT: 4   QUERY(truncated): SELECT DISTINCT t0.ID, t0.DTYPE, t0.DESCRIPTION, t0.DISPLAYNAME, t0.PERSISTEDGROUPALIAS FROM IPV4RANGE t1 LEFT OUTER JOIN PERSISTEDGLOBALGROUP t0 ON (t0.ID = t1.OWNER_ID) WHERE ((t1.BOTTOMASLONG <= ?) AND (t1.TOPASLONG >= ?))
```

**Comment by bencomp** [comment]
I'm pretty sure the following in `filesFragment.xhtml` is causing the server events:
```
<p:ajax event="rowSelectCheckbox"
        listener="#{DatasetPage.updateFileCounts}"
        process="filesTable"
        update="zipDownloadPanel">
</p:ajax>
<p:ajax event="rowUnselectCheckbox"
        listener="#{DatasetPage.updateFileCounts}"
        update="zipDownloadPanel">
</p:ajax>
```
The `DatasetPage.updateFileCounts()` method checks for all selected files whether it is restricted and whether the file is a tabular file. I suppose this 'real-time' check for updates on file restrictions is useful... err, when several people are editing and viewing the (by definition) draft dataset version simultaneously. I'm pretty sure this is a use case that very few people need.

**Comment by raprasad** [comment]
Thanks @bencomp 
We were also looking at the 51/64 queries in the non-logged in state just from this:
```xhtml
<p:ajax event="rowSelectCheckbox">
</p:ajax>
```
This is after removing the ```listener```, ```process```, and ```update``` parameters from ```<p:ajax```:
```
listener="#{DatasetPage.updateFileCounts}"
process="filesTable"
update="zipDownloadPanel"
```

**Comment by eaquigley** [comment]
@raprasad @sekmiller @scolapasta Would it make sense to do a code review for this while @sekmiller works on this? it would be great if @bencomp is available and willing to chime in on that code review (no pressure though).

**Comment by pdurbin** [comment]
@raprasad there has been a lot of activity on the dataset page for 4.2.1. Do you know if this is any better? Does @sekmiller know? Should this issue really be in the "Candidates for 4.4 milestone"?

**Comment by pdurbin** [comment]
I checked with @sekmiller about this and he said it should be fixed in the 4.2.1 branch so I'm setting that as the milestone and sending this to QA.

**Comment by kcondon** [comment]
I am seeing no net db counts in logs when checking checkbox. Closing.
